### ISSUE TITLE
[FR] - Provide high quality multiarch docker images

### ISSUE THREAD

**Issue by tdiesler** [input-output-hk/cardano-node#2360]
**Internal/External**
*External* 

**Area**
*Docker/Kubernetes*   

**Describe the feature you'd like**

This is an umbrella issue that combines various feature requests and bug fixes in order to provide a high quality Cardano Docker image. In the not too distant future it may become a reality that Cardano nodes need to get integrated in corporate  IT infrastructure. They may become destinations or sources for various corporate data streams and as such, it'd be good to have an offering that allows these organizations to integrate Cardano with already existing cloud infrastructure. 

With Kubernetes as the de-facto standard for container deployment, orchestration, monitoring, scaling , etc, it should be as easy as this to integrate a Cardano node ...

```
kubectl apply -f https://raw.githubusercontent.com/input-output-hk/cardano-node/master/cardano-node.yaml
```

This issue suggests the following improvements ...

* [#2361] Added support for arm64 and perhaps for other target platforms to follow
* [#2353] Support for graceful shutdown in container environments
* [#2362] Self contained topology updates (until there is native p2p support provided by the node)
* [#2364] Comprehensive documentation on how to run Cardano in managed container environments

**Describe alternatives you've considered**

An alternative would be to have advanced image functionality be provided by 3rd party and let this repository continue to focus on bare metal installations and more or less leave the existing (x86_64) docker image as is.

IMHO, it'd however be much better to have this tech be provided upstream.

**Additional context / screenshots**

As far as I can tell, much of the above is low hanging fruit and if it's ok with you, I'd volunteer to do this over the course of the following weeks/months. The multiarch bit may initially require a workaround until we have the regular nix build that works on arm64. The image cannot rely on an external process to do topology updates - this has to be done within. Initially we could perhaps use guild-operators for that (to be discussed). Kubernetes support and getting started docs should not be a big deal.

**Comment by Preskton** [comment]
Hi, @tdiesler - I'd also offer up https://github.com/input-output-hk/cardano-node/issues/2423 as an optimization after we work through some of your above list.

**Comment by tdiesler** [comment]
Status ...

PR [#2456] is the fivth and last installment of [#2360] Provide high quality multiarch docker image and k8s support

Now we have ...

* Preserved functionality via -e NETWORK
* Node configuration through env vars, which is later needed for k8s integration
* Deployment scripts for Kubernetes and Docker Compose
* Integrated configurable topology update
* Graceful shutdown on docker stop
* Improved access to the CLI
* Comprehensive documentation for Kubernetes & Compose

There is a README about this [here](https://github.com/tdiesler/cardano-node/tree/docker/nix/docker).

PS: Once this is merged we'd need to cleanup a few links and perhaps find a better home for the docs.

**Comment by Santander2020** [comment]
Sign up for the first Bitcoin rewards card and get 5,000 sats for using my
referral link to sign up for Fold! https://use.foldapp.com/r/XRNKXFHY

Вс, 7 марта 2021 г. в 19:32, Thomas Diesler <notifications@github.com>:

**Comment by coot** [comment]
Thanks for raising this issue.  We're certainly fine with providing good docker story, but we strongly discourage using docker based deployment platforms like kubernetes or ecs.  These platforms, though popular in deploying various system on the internet, are not suitable for `cardano` for several reasons which I highlight below:

* `cardano` needs a highly secure environment, where the access to private keys is ensured to be secure; These deployment platforms give access to private keys by the hosting company which can compromise a node. Not only stealing is at stake, but also acting as an adversary to the rest of the system.
* `cardano` is a real time system with real time constraints; this needs to be taken into consideration seriously otherwise it might not meet its deadlines.  Unfortunately k8s or ecs are not suitable for real time systems.  To our knowledge this platforms do not take into consideration scaling of such systems.
* managing large data sets on which cardano operates (blockchain db), is a non trivial task. As Google points out (in relation to services such as databases) this requires careful consideration and interacts with management and resilience issues.

Providing good docker support should come together with a good documentation that points out the potential threats one needs to be aware when deploying docker images. Real time distributed systems with very strict security model running on the world scale are very rare so we need to be quite careful about deployment platforms which were developed for applications that use a very different model.

**Comment by tdiesler** [comment]
Thanks for pointing this out @coot. This PR does actually not include the K8S stuff that the incubator is still experimenting with (I changed the title of the PR to reflect that). I'd also like to see how it behaves with the Alonzo P2P manager before even proposing this upstream. What you see here mainly improves the existing docker image in a number of areas, namely

* graceful shutdown
* config through env vars
* builtin topology updates
* minor bugs

If you prefer, we can take the docker compose example script out as well, because relay and block producer are supposed to run on different hosts anyway.

In the long run however, I still think that the node should have a decent docker image, which ideally should support multiple platforms. I agree, the Docker story needs to be well documented - much better than it is now.

**Comment by coot** [comment]
> If you prefer, we can take the docker compose example script out as well, because relay and block producer are supposed to run on different hosts anyway.

Yes, running both on the same virtual machine is not a good idea.

Also, once peer2peer will be released, topology updates will not be necessary.

**Comment by tdiesler** [comment]
Ok, this is currently being discussed on #2608. Perhaps you'd like to make yourself heard over there as well. I guess it's in all of our interest to only get stuff upstream that we really want.

Specifically ...

* docker compose in/out
* background topology updates in/out

IIUYC, you'd opt for both out, right?

**Comment by coot** [comment]
only docker compose; topology updater is ok.  We will be slowly merging p2p into `master`, which eventually will be enabled with an experimental flag.  p2p-nodes will not require topology updater, and will use a different format for topology file.

**Comment by tdiesler** [comment]
Rebased PR: #2802

**Comment by tdiesler** [comment]
Most of what we wanted to do is now upstream. Docker Compose and Kubernetes stuff has been dropped. We also decided to postpone the script based topology update for now and revisit if Alonzo p2p should get delayed for some reason.

**Comment by aldegoeij** [comment]
Like to give some rebutt to @coot 's points on why no containers, I think they might be an oversimplification:

> * `cardano` needs a highly secure environment, where the access to private keys is ensured to be secure; These deployment platforms give access to private keys by the hosting company which can compromise a node. Not only stealing is at stake, but also acting as an adversary to the rest of the system.

I don't think there is any relation between the tech you use and the level of security, running on e.g. AWS there is no difference in level of security whether you are running on an EC2 virtual machine or in a ECS container, as a matter of fact a container is more secure as the options to access it via any other means than the load balancer (a secured endpoint) are extremely limited, e.g. you won't find a forgotton SSH port open on an ECS container.

Cardano docs state very clearly that cold keys should not be available on running nodes and only required on startup, something that can be easily handled with the standard secret manager tools on ECS or Kubernetes (EKS)

> * `cardano` is a real time system with real time constraints; this needs to be taken into consideration seriously otherwise it might not meet its deadlines.  Unfortunately k8s or ecs are not suitable for real time systems.  To our knowledge this platforms do not take into consideration scaling of such systems.

Not sure why ECS or K8S or any container provider has any issues with real-time behaviour, as a matter of fact, most of the worlds most real-time services run on k8s. As a matter of fact the elasticity of containers provides a much more robust and scalable infra than depending on EC2 vms which take a long time to start and are long-lived in nature.

> * managing large data sets on which cardano operates (blockchain db), is a non trivial task. As Google points out (in relation to services such as databases) this requires careful consideration and interacts with management and resilience issues.

Not sure I understand how this point relates to containers.

> Providing good docker support should come together with a good documentation that points out the potential threats one needs to be aware when deploying docker images. Real time distributed systems with very strict security model running on the world scale are very rare so we need to be quite careful about deployment platforms which were developed for applications that use a very different model.

Very much agree a deployment option should be properly documented, and that is I think exactly the reason tickets like this one are here.

Reading through architecture documentation and what others have posted on their setup, it seems easy to run Relay nodes on container platforms, as a matter of fact they could easily scale up and down with the amount of traffic hitting them, however the design of the block producing node (IMHO) is not really suited for containers.

The BP node operates on a full copy of the chain (currently 8gb+) which means starting from scratch takes a long (<24h) time, which is unacceptable in a HA environment.

Also it does not seem the BP node's signing of blocks is something that would work OOTB with multiple writers to a single dataset (in the end there is only one write to be done, the new block creation).

So agreed that in the current state running multiple BP node _containers_ for a single pool makes little sense.

So how does it work for the BP node, does it have to process _all_ transactions on the network to be possible elected to be leader? Or is there sharding of nodes, otherwise BP node compute capacity has to scale liniarly with the transaction volume?

**Comment by tdiesler** [comment]
@aldegoeij thanks for your feedback. We can still use the [incubator project](https://github.com/tdiesler/nessus-cardano/tree/master/nix/docker/k8s) to experiment and explore management/orchestration for cardano nodes. For now, I'm happy that the essentials found their way into the upstream codebase.

**Comment by marcellodesales** [comment]
I have a plan: run a huge Kubernetes cluster with this work! :) Containers are gold in any cloud, more secure than anything else (if you know what you are doing...)... I agree with the points described by @aldegoeij... I ended up here after looking for cardano docker :) Left the slavery of a 9-5 to a libertarian world of blockchain engineering :D
